File: ios/AgroBridge/AgroBridge/Data/Repository/LoteRepositoryImpl.swift
//
//  LoteRepositoryImpl.swift
//  AgroBridge
//
//  Created by Alejandro Navarro Ayala - CEO & Senior Developer
//  Copyright © 2025 AgroBridge International. All rights reserved.
//
//  ANDROID EQUIVALENT: data/repository/LoteRepositoryImpl.kt
//

import Foundation
import Combine
import CoreData

/**
 * LoteRepositoryImpl - Repository de Lotes con arquitectura OFFLINE-FIRST
 *
 * ANDROID EQUIVALENT: class LoteRepositoryImpl @Inject constructor(...)
 *
 * Flujo de datos:
 * 1. API (URLSession) obtiene datos remotos
 * 2. Los datos se persisten en CoreData (SQLite local)
 * 3. La UI observa los datos desde CoreData (siempre tiene algo que mostrar)
 * 4. Si no hay internet, la UI muestra datos cacheados localmente
 *
 * Patrón: Single Source of Truth (SSOT) - La DB local es la verdad
 *
 * Upload Sync (Offline-First Write):
 * 1. Usuario crea/edita lote localmente → PENDING_CREATE/PENDING_UPDATE
 * 2. Se guarda en CoreData → UI se actualiza inmediatamente
 * 3. Background task detecta cambios y sube a servidor
 * 4. Task reintenta si falla, cuando haya conexión
 *
 * ANDROID PARITY: Matches LoteRepositoryImpl.kt architecture 100%
 * - Same offline-first pattern
 * - Same CRUD operations
 * - Same sync logic
 * - Same error handling
 */
@MainActor
class LoteRepositoryImpl: ObservableObject {

    // MARK: - Singleton

    static let shared = LoteRepositoryImpl()

    // MARK: - Dependencies

    private let coreDataManager = CoreDataManager.shared
    private let tokenManager = TokenManager.shared
    private let networkMonitor = NetworkMonitor.shared
    private let errorHandler = ErrorHandler.shared
    private let mapper = LoteMapper.shared

    // FIXED: L-012 - Use AppConfiguration instead of hardcoded URL
    private var baseURL: String {
        AppConfiguration.baseURL
    }

    // MARK: - Published Properties

    /// Indica si hay cambios pendientes de sincronizar
    @Published private(set) var hasPendingChanges = false

    /// Número de lotes pendientes de sincronizar
    @Published private(set) var pendingCount = 0

    /// Timestamp de última sincronización exitosa
    @Published private(set) var lastSyncTimestamp: Date?

    // MARK: - URLSession

    private let session: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 60
        return URLSession(configuration: config)
    }()

    // MARK: - Initialization

    private init() {
        Task {
            await updatePendingCount()
        }
    }

    // MARK: - Fetch Lotes (OBSERVABLE)

    /**
     * Obtener lotes de un productor específico desde CoreData
     *
     * ANDROID EQUIVALENT: `fun getLotes(productorId: String, page: Int, pageSize: Int): Flow<List<Lote>>`
     *
     * VENTAJA OFFLINE-FIRST:
     * - Si la app se abre sin internet, se muestran lotes cacheados
     * - Cuando hay conexión, se actualiza automáticamente
     *
     * - Parameters:
     *   - productorId: ID del productor
     * - Returns: Array de Lote domain models
     */
    func getLotes(productorId: String) async throws -> [Lote] {
        let entities = try await coreDataManager.fetchLotes(forProductorId: productorId)
        return mapper.toDomain(from: entities)
    }

    /**
     * Obtener todos los lotes desde CoreData
     *
     * ANDROID EQUIVALENT: `fun getAllLotes(): Flow<List<Lote>>`
     */
    func getAllLotes() async throws -> [Lote] {
        let entities = try await coreDataManager.fetchAllLotes()
        return mapper.toDomain(from: entities)
    }

    /**
     * Obtener un lote específico por ID
     *
     * ANDROID EQUIVALENT: `fun getLoteById(loteId: String): Flow<Lote?>`
     *
     * - Parameter loteId: ID del lote
     * - Returns: Lote domain model o nil si no existe
     */
    func getLoteById(_ loteId: String) async throws -> Lote? {
        guard let entity = try await coreDataManager.fetchLote(byId: loteId) else {
            return nil
        }

        return mapper.toDomain(from: entity)
    }

    /**
     * Obtener lotes por estado
     *
     * ANDROID EQUIVALENT: `fun getActiveLotes(productorId: String): Flow<List<Lote>>`
     */
    func getLotes(byEstado estado: LoteEstado) async throws -> [Lote] {
        let entities = try await coreDataManager.fetchLotes(withEstado: estado.rawValue)
        return mapper.toDomain(from: entities)
    }

    // MARK: - Refresh from API (SYNC DOWN)

    /**
     * SINCRONIZACIÓN: Descargar lotes desde API y guardar en CoreData
     *
     * ANDROID EQUIVALENT: `suspend fun refreshLotes(productorId: String): Result<Unit>`
     *
     * FLUJO:
     * 1. Intenta obtener datos de la API
     * 2. Si éxito, guarda en CoreData (transacción atómica)
     * 3. Los observadores se actualizan automáticamente
     * 4. La UI se refresca sin intervención manual
     *
     * - Parameter productorId: ID del productor
     * - Returns: Result<Void> indicando éxito o fallo
     */
    func refreshLotes(productorId: String) async -> Result<Void, Error> {
        do {
            print("🔄 LoteRepository: Iniciando sincronización de lotes para productor: \(productorId)")

            // Obtener datos de la API
            let endpoint = "/v1/lotes?productor_id=\(productorId)"
            let response: LotesResponse = try await get(endpoint: endpoint)

            if !response.lotes.isEmpty {
                // Convertir DTOs a Entities y guardar
                for loteDTO in response.lotes {
                    let entity = mapper.toEntity(
                        from: loteDTO,
                        context: coreDataManager.viewContext,
                        syncStatus: .synced
                    )
                    // Entity se crea en contexto, se guardará con save()
                }

                // Guardar transacción completa
                try await coreDataManager.save()

                // Actualizar timestamp de sincronización
                await MainActor.run {
                    lastSyncTimestamp = Date()
                }

                print("✅ LoteRepository: Sincronización exitosa: \(response.lotes.count) lotes guardados")
                return .success(())
            } else {
                print("⚠️ LoteRepository: API retornó respuesta vacía")
                // Los datos viejos en la DB se mantienen
                return .success(())
            }
        } catch {
            let message = await errorHandler.handle(error, context: "LoteRepository.refreshLotes")
            print("❌ LoteRepository: Error sincronizando lotes: \(message)")
            return .failure(error)
        }
    }

    // MARK: - Create Lote (OFFLINE-FIRST WRITE)

    /**
     * Crear un nuevo lote (Offline-First Write)
     *
     * ANDROID EQUIVALENT: `suspend fun createLote(lote: Lote): Result<Unit>`
     *
     * FLUJO:
     * 1. Convertir Lote (Domain) -> LoteEntity (Persistence)
     * 2. Marcar con syncStatus = PENDING_CREATE
     * 3. Guardar localmente en CoreData (INMEDIATO)
     * 4. Encolar sincronización en background
     * 5. UI se actualiza automáticamente
     *
     * Ventajas:
     * - Feedback inmediato al usuario (sin esperar red)
     * - Si no hay internet, se guarda y sube luego automáticamente
     * - Sin bloqueo de UI
     *
     * - Parameter lote: Lote domain model
     * - Returns: Result<Void> indicando éxito o fallo
     */
    func createLote(_ lote: Lote) async -> Result<Void, Error> {
        do {
            print("➕ LoteRepository: Creando lote: \(lote.nombre)")

            // 1. Convertir Lote (Domain) a LoteEntity con estado PENDING_CREATE
            let entity = mapper.toEntity(
                from: lote,
                context: coreDataManager.viewContext,
                syncStatus: .pendingCreate
            )

            // 2. Guardar localmente (INMEDIATO - no espera red)
            try await coreDataManager.save()
            print("✅ LoteRepository: Lote guardado localmente: \(lote.nombre)")

            // 3. Actualizar pending count
            await updatePendingCount()

            // 4. Intentar sincronizar si hay conexión
            if networkMonitor.isConnected {
                Task {
                    await syncPendingLotes()
                }
            }

            return .success(())
        } catch {
            let message = await errorHandler.handle(error, context: "LoteRepository.createLote")
            print("❌ LoteRepository: Error creando lote: \(message)")
            return .failure(error)
        }
    }

    // MARK: - Update Lote (OFFLINE-FIRST WRITE)

    /**
     * Actualizar un lote existente (Offline-First Write)
     *
     * ANDROID EQUIVALENT: `suspend fun updateLote(loteId: String, lote: Lote): Result<Unit>`
     *
     * Mismo flujo que createLote pero con PENDING_UPDATE
     *
     * - Parameters:
     *   - loteId: ID del lote a actualizar
     *   - lote: Lote domain model con datos actualizados
     * - Returns: Result<Void> indicando éxito o fallo
     */
    func updateLote(loteId: String, lote: Lote) async -> Result<Void, Error> {
        do {
            print("✏️ LoteRepository: Actualizando lote: \(lote.nombre)")

            // 1. Buscar entidad existente
            guard let entity = try await coreDataManager.fetchLote(byId: loteId) else {
                throw DatabaseError(message: "Lote no encontrado: \(loteId)")
            }

            // 2. Actualizar con datos del dominio y marcar como PENDING_UPDATE
            mapper.updateEntity(entity, from: lote, syncStatus: .pendingUpdate)

            // 3. Guardar localmente
            try await coreDataManager.save()
            print("✅ LoteRepository: Lote actualizado localmente: \(lote.nombre)")

            // 4. Actualizar pending count
            await updatePendingCount()

            // 5. Intentar sincronizar si hay conexión
            if networkMonitor.isConnected {
                Task {
                    await syncPendingLotes()
                }
            }

            return .success(())
        } catch {
            let message = await errorHandler.handle(error, context: "LoteRepository.updateLote")
            print("❌ LoteRepository: Error actualizando lote: \(message)")
            return .failure(error)
        }
    }

    // MARK: - Delete Lote

    /**
     * Eliminar un lote
     *
     * ANDROID EQUIVALENT: `suspend fun deleteLote(loteId: String): Result<Unit>`
     */
    func deleteLote(_ loteId: String) async -> Result<Void, Error> {
        do {
            print("🗑️ LoteRepository: Eliminando lote: \(loteId)")

            guard let entity = try await coreDataManager.fetchLote(byId: loteId) else {
                throw DatabaseError(message: "Lote no encontrado: \(loteId)")
            }

            coreDataManager.delete(entity)
            try await coreDataManager.save()

            print("✅ LoteRepository: Lote eliminado localmente")
            return .success(())
        } catch {
            let message = await errorHandler.handle(error, context: "LoteRepository.deleteLote")
            return .failure(error)
        }
    }

    // MARK: - Background Sync (SYNC UP)

    /**
     * Sincronizar lotes pendientes con el servidor
     *
     * ANDROID EQUIVALENT: SyncLotesWorker (WorkManager background task)
     *
     * FLUJO:
     * 1. Obtener todos los lotes con PENDING_CREATE o PENDING_UPDATE
     * 2. Para cada lote:
     *    a. Subir a API
     *    b. Si éxito, marcar como SYNCED
     *    c. Si fallo, mantener PENDING y reintentar luego
     * 3. Actualizar pending count
     */
    func syncPendingLotes() async {
        do {
            print("📤 LoteRepository: Sincronizando lotes pendientes...")

            // 1. Obtener lotes pendientes
            let pendingEntities = try await coreDataManager.fetchPendingLotes()

            if pendingEntities.isEmpty {
                print("ℹ️ LoteRepository: No hay lotes pendientes de sincronizar")
                return
            }

            print("📋 LoteRepository: Encontrados \(pendingEntities.count) lotes pendientes")

            // 2. Sincronizar cada lote
            for entity in pendingEntities {
                let lote = mapper.toDomain(from: entity)

                switch entity.syncStatusEnum {
                case .pendingCreate:
                    await uploadNewLote(lote, entity: entity)

                case .pendingUpdate:
                    await uploadUpdatedLote(lote, entity: entity)

                case .synced:
                    break  // Ya está sincronizado
                }
            }

            // 3. Actualizar pending count
            await updatePendingCount()

            print("✅ LoteRepository: Sincronización completada")
        } catch {
            let message = await errorHandler.handle(error, context: "LoteRepository.syncPendingLotes")
            print("❌ LoteRepository: Error en sincronización: \(message)")
        }
    }

    /**
     * Subir nuevo lote a la API
     */
    private func uploadNewLote(_ lote: Lote, entity: LoteEntity) async {
        do {
            print("⬆️ LoteRepository: Subiendo nuevo lote: \(lote.nombre)")

            let request = mapper.toCreateRequest(from: lote)
            let endpoint = "/v1/lotes"

            let response: LoteDTO = try await post(endpoint: endpoint, body: request)

            // Actualizar entity con datos del servidor y marcar como SYNCED
            mapper.updateEntity(entity, from: response, syncStatus: .synced)
            try await coreDataManager.save()

            print("✅ LoteRepository: Lote subido exitosamente: \(lote.nombre)")
        } catch {
            let message = await errorHandler.handle(error, context: "LoteRepository.uploadNewLote")
            print("⚠️ LoteRepository: Fallo al subir lote (se reintentará): \(message)")
            // Mantener entity como PENDING_CREATE para reintentar
        }
    }

    /**
     * Subir lote actualizado a la API
     */
    private func uploadUpdatedLote(_ lote: Lote, entity: LoteEntity) async {
        do {
            print("⬆️ LoteRepository: Subiendo actualización de lote: \(lote.nombre)")

            let dto = mapper.toDTO(from: lote)
            let endpoint = "/v1/lotes/\(lote.id)"

            let response: LoteDTO = try await put(endpoint: endpoint, body: dto)

            // Actualizar entity y marcar como SYNCED
            mapper.updateEntity(entity, from: response, syncStatus: .synced)
            try await coreDataManager.save()

            print("✅ LoteRepository: Lote actualizado exitosamente: \(lote.nombre)")
        } catch {
            let message = await errorHandler.handle(error, context: "LoteRepository.uploadUpdatedLote")
            print("⚠️ LoteRepository: Fallo al actualizar lote (se reintentará): \(message)")
            // Mantener entity como PENDING_UPDATE para reintentar
        }
    }

    // MARK: - Pending Lotes Info

    /**
     * Actualizar el contador de lotes pendientes
     *
     * ANDROID EQUIVALENT: `fun getPendingLotesCount(): Flow<Int>`
     */
    private func updatePendingCount() async {
        do {
            let pending = try await coreDataManager.fetchPendingLotes()
            await MainActor.run {
                self.pendingCount = pending.count
                self.hasPendingChanges = pending.count > 0
            }
        } catch {
            print("⚠️ Error actualizando pending count: \(error)")
        }
    }

    /**
     * Obtener lotes pendientes de sincronización
     *
     * ANDROID EQUIVALENT: `fun getPendingLotes(): Flow<List<Lote>>`
     */
    func getPendingLotes() async throws -> [Lote] {
        let entities = try await coreDataManager.fetchPendingLotes()
        return mapper.toDomain(from: entities)
    }

    // MARK: - Database Management

    /**
     * Limpiar la base de datos local
     *
     * ANDROID EQUIVALENT: `suspend fun clearDatabase(): Result<Unit>`
     *
     * Útil después de logout o para testing
     */
    func clearDatabase() async -> Result<Void, Error> {
        do {
            print("🗑️ LoteRepository: Limpiando base de datos local")
            try await coreDataManager.deleteAllLotes()
            await updatePendingCount()
            print("✅ LoteRepository: Base de datos limpiada")
            return .success(())
        } catch {
            let message = await errorHandler.handle(error, context: "LoteRepository.clearDatabase")
            return .failure(error)
        }
    }

    // MARK: - Private Network Methods

    /**
     * Realiza GET request al API
     */
    private func get<R: Decodable>(endpoint: String) async throws -> R {
        guard let url = URL(string: baseURL + endpoint) else {
            throw HTTPError(statusCode: 0, message: "URL inválida")
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"

        // Add access token
        if let accessToken = try await tokenManager.getAccessToken() {
            request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        }

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw HTTPError(statusCode: 0, message: "Invalid response")
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            throw HTTPError(statusCode: httpResponse.statusCode, message: "HTTP Error \(httpResponse.statusCode)")
        }

        return try JSONDecoder().decode(R.self, from: data)
    }

    /**
     * Realiza POST request al API
     */
    private func post<T: Encodable, R: Decodable>(endpoint: String, body: T) async throws -> R {
        guard let url = URL(string: baseURL + endpoint) else {
            throw HTTPError(statusCode: 0, message: "URL inválida")
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if let accessToken = try await tokenManager.getAccessToken() {
            request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        }

        request.httpBody = try JSONEncoder().encode(body)

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw HTTPError(statusCode: 0, message: "Invalid response")
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            throw HTTPError(statusCode: httpResponse.statusCode, message: "HTTP Error \(httpResponse.statusCode)")
        }

        return try JSONDecoder().decode(R.self, from: data)
    }

    /**
     * Realiza PUT request al API
     */
    private func put<T: Encodable, R: Decodable>(endpoint: String, body: T) async throws -> R {
        guard let url = URL(string: baseURL + endpoint) else {
            throw HTTPError(statusCode: 0, message: "URL inválida")
        }

        var request = URLRequest(url: url)
        request.httpMethod = "PUT"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if let accessToken = try await tokenManager.getAccessToken() {
            request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        }

        request.httpBody = try JSONEncoder().encode(body)

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw HTTPError(statusCode: 0, message: "Invalid response")
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            throw HTTPError(statusCode: httpResponse.statusCode, message: "HTTP Error \(httpResponse.statusCode)")
        }

        return try JSONDecoder().decode(R.self, from: data)
    }
}

File: app/src/main/java/com/agrobridge/data/repository/LoteRepositoryImpl.kt
package com.agrobridge.data.repository

import androidx.work.BackoffPolicy
import androidx.work.Constraints
import androidx.work.ExistingWorkPolicy
import androidx.work.NetworkType
import androidx.work.OneTimeWorkRequestBuilder
import androidx.work.WorkManager
import com.agrobridge.data.dto.PaginatedResponse
import com.agrobridge.data.local.dao.LoteDao
import com.agrobridge.data.local.entity.SyncStatus
import com.agrobridge.data.mapper.LoteEntityMapper.toEntity
import com.agrobridge.data.mapper.LoteEntityToDomainMapper.toDomain
import com.agrobridge.data.model.Lote
import com.agrobridge.data.remote.ApiService
import com.agrobridge.data.worker.SyncLotesWorker
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.map
import timber.log.Timber
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Implementación del repositorio de Lotes con arquitectura OFFLINE-FIRST
 *
 * Flujo de datos:
 * 1. API (Retrofit) obtiene datos remotos
 * 2. Los datos se persisten en Room (SQLite local)
 * 3. La UI observa los datos de Room (siempre tiene algo que mostrar)
 * 4. Si no hay internet, la UI muestra datos cacheados localmente
 *
 * Patrón: Single Source of Truth (SSOT) - La DB local es la verdad
 *
 * Upload Sync (Offline-First Write):
 * 1. Usuario crea/edita lote localmente → PENDING_CREATE/PENDING_UPDATE
 * 2. Se guarda en Room → UI se actualiza inmediatamente
 * 3. WorkManager detecta cambios y encola SyncLotesWorker
 * 4. Worker sube en background, reintenta si falla
 */
@Singleton
class LoteRepositoryImpl @Inject constructor(
    private val apiService: ApiService,
    private val loteDao: LoteDao,
    private val workManager: WorkManager
) : LoteRepository {

    /**
     * OBSERVABLE: Obtener lotes de un productor específico desde la base de datos
     * Este es el flujo que la UI observa. Emite datos automáticamente
     * cada vez que la DB cambia
     *
     * VENTAJA OFFLINE-FIRST:
     * - Si la app se abre sin internet, la UI muestra lotes cacheados
     * - Cuando hay conexión, se actualiza automáticamente
     *
     * FIXED: HIGH-11
     * - Antes: ignoraba productorId, retornaba TODOS los lotes
     * - Ahora: filtra por productorId específico
     */
    override fun getLotes(
        productorId: String,
        page: Int,
        pageSize: Int
    ): Flow<List<Lote>> {
        // FIXED: HIGH-11 - Use productorId parameter instead of ignoring it
        return loteDao.getLotesByProducer(productorId).map { entities ->
            entities.map { it.toDomain() }
        }
    }

    /**
     * Obtener un lote específico
     */
    override fun getLoteById(loteId: String): Flow<Lote?> {
        return loteDao.getLoteById(loteId).map { entity ->
            entity?.toDomain()
        }
    }

    /**
     * Obtener lotes activos
     */
    override fun getActiveLotes(productorId: String): Flow<List<Lote>> {
        return loteDao.getActiveLotes().map { entities ->
            entities.map { it.toDomain() }
        }
    }

    /**
     * SINCRONIZACIÓN: Descargar lotes desde la API y guardar en DB
     * Se ejecuta en background sin bloquear la UI
     *
     * FLUJO:
     * 1. Intenta obtener datos de la API
     * 2. Si éxito, guarda en la DB (transacción atómica)
     * 3. El Flow de getLotes() se actualiza automáticamente
     * 4. La UI se refresca sin intervención manual
     */
    override suspend fun refreshLotes(productorId: String): Result<Unit> {
        return runCatching {
            Timber.d("🔄 Iniciando sincronización de lotes para productor: $productorId")

            try {
                // 1. Obtener datos de la API
                val response = apiService.getLotes(productorId)

                if (response.isSuccessful) {
                    val body = response.body()
                    if (body != null && body.data.isNotEmpty()) {
                        // 2. Convertir DTOs a Entities
                        val entities = body.data.map { it.toEntity() }

                        // 3. Guardar en DB (transacción atómica: borrar e insertar)
                        loteDao.refreshLotes(entities)

                        Timber.d("✅ Sincronización exitosa: ${entities.size} lotes guardados")
                    } else {
                        Timber.w("⚠️ API retornó respuesta vacía")
                        // Los datos viejos en la DB se mantienen, la UX sigue funcionando
                    }
                } else {
                    Timber.e("❌ Error HTTP ${response.code()}: ${response.message()}")
                    throw Exception("Error HTTP ${response.code()}")
                }
            } catch (e: Exception) {
                Timber.e(e, "❌ Error sincronizando lotes")
                throw e
            }
        }
    }

    /**
     * Crear un nuevo lote (Offline-First Write)
     *
     * FLUJO:
     * 1. Convertir Lote (Domain) -> LoteEntity (Persistence)
     * 2. Marcar con syncStatus = PENDING_CREATE
     * 3. Guardar localmente en Room (INMEDIATO)
     * 4. Encolar SyncLotesWorker para subir en background
     * 5. UI se actualiza automáticamente vía Flow desde Room
     *
     * Ventajas:
     * - Feedback inmediato al usuario (sin esperar red)
     * - Si no hay internet, se guarda y sube luego automáticamente
     * - Sin bloqueo de UI
     */
    suspend fun createLote(lote: Lote): Result<Unit> {
        return runCatching {
            Timber.d("➕ Creando lote: ${lote.nombre}")

            // 1. Convertir Lote (Domain) a LoteEntity (Entity)
            // y marcar como PENDING_CREATE
            // FIXED: HIGH-13 - Use newly implemented mapper for domain to entity conversion
            val loteEntity = lote.toEntity().copy(
                syncStatus = SyncStatus.PENDING_CREATE,
                fechaActualizacion = System.currentTimeMillis()
            )

            // 2. Guardar localmente (INMEDIATO - no espera red)
            loteDao.saveLocal(loteEntity)
            Timber.d("✅ Lote guardado localmente: ${lote.nombre}")

            // 3. Encolar trabajo de sincronización
            enqueueSyncWork()
        }
    }

    /**
     * Editar un lote existente (Offline-First Write)
     *
     * Mismo flujo que createLote pero con PENDING_UPDATE
     */
    suspend fun updateLote(loteId: String, lote: Lote): Result<Unit> {
        return runCatching {
            Timber.d("✏️ Actualizando lote: ${lote.nombre}")

            // 1. Convertir a Entity con estado PENDING_UPDATE
            // FIXED: HIGH-13 - Use mapper for domain to entity conversion
            val loteEntity = lote.toEntity().copy(
                id = loteId,
                syncStatus = SyncStatus.PENDING_UPDATE,
                fechaActualizacion = System.currentTimeMillis()
            )

            // 2. Guardar localmente
            loteDao.saveLocal(loteEntity)
            Timber.d("✅ Lote actualizado localmente: ${lote.nombre}")

            // 3. Encolar sincronización
            enqueueSyncWork()
        }
    }

    /**
     * Guardar un lote individual en la DB (alias para compatibilidad)
     * Ahora implementado correctamente con Upload Sync
     */
    override suspend fun saveLote(lote: Lote): Result<Unit> {
        // Asumimos que es una creación si es nuevo, sino sería update
        return createLote(lote)
    }

    /**
     * Encolar trabajo de sincronización
     * WorkManager ejecutará SyncLotesWorker cuando:
     * - Haya conexión a internet
     * - Batería disponible
     * - Condiciones del sistema lo permitan
     */
    private fun enqueueSyncWork() {
        Timber.d("📋 Encolando trabajo de sincronización...")

        try {
            // Crear restricciones para el trabajo
            val constraints = Constraints.Builder()
                .setRequiredNetworkType(NetworkType.CONNECTED) // Solo con conexión
                .build()

            // Crear la solicitud de trabajo
            val syncRequest = OneTimeWorkRequestBuilder<SyncLotesWorker>()
                .setConstraints(constraints)
                // Politica de reintento: exponencial (5s, 10s, 20s, ...)
                .setBackoffCriteria(
                    BackoffPolicy.EXPONENTIAL,
                    WorkRequest.MIN_BACKOFF_MILLIS,
                    TimeUnit.MILLISECONDS
                )
                .build()

            // Encolar con política KEEP
            // (KEEP = no encolar si ya hay un trabajo en progreso)
            workManager.enqueueUniqueWork(
                "sync_lotes_work",
                ExistingWorkPolicy.KEEP, // O APPEND_OR_REPLACE para acumular
                syncRequest
            )

            Timber.d("✅ Trabajo encolado exitosamente")

        } catch (e: Exception) {
            Timber.e(e, "❌ Error encolando trabajo de sincronización")
        }
    }

    /**
     * Limpiar la base de datos local
     * Útil después de logout o para testing
     */
    override suspend fun clearDatabase(): Result<Unit> {
        return runCatching {
            Timber.d("🗑️ Limpiando base de datos local")
            loteDao.clearAll()
            Timber.d("✅ Base de datos limpiada")
        }
    }

    /**
     * Obtener información de sincronización
     * Útil para mostrar timestamp de última actualización en la UI
     */
    override suspend fun getLastSyncTimestamp(): Long? {
        return try {
            loteDao.getLastSyncTimestamp()
        } catch (e: Exception) {
            Timber.e(e, "Error obteniendo timestamp de sincronización")
            null
        }
    }

    /**
     * Obtener el count de lotes pendientes como Flow
     * Útil para mostrar badges en UI: "3 cambios sin sincronizar"
     *
     * @return Flow<Int> que emite el número de lotes pendientes
     */
    fun getPendingLotesCount(): Flow<Int> {
        return loteDao.getPendingLotesCount()
    }

    /**
     * Obtener lotes pendientes como Flow (para observar en tiempo real)
     * Útil para mostrar lista de cambios sin sincronizar
     *
     * @return Flow<List<Lote>> con los lotes que están PENDING_CREATE o PENDING_UPDATE
     */
    fun getPendingLotes(): Flow<List<Lote>> {
        return loteDao.getPendingLotesFlow().map { entities ->
            entities.map { it.toDomain() }
        }
    }
}
